lib/operations/crud_write.js

```diff
@@ -0,0 +1,282 @@
+  constructor(model) {
+    super(model.options);
+    this.model = model;
+  }
+
+  canRetry(server, topology) {
+    return (
+      this.options.retryWrites &&
+      this.session &&
+      !this.session.inTransaction() &&
+      isRetryabilitySupported(topology) &&
+      this.model.canRetry()
+    );
+  }
+
+  enableRetry() {
+    this.model.options.willRetryWrite = true;
+    this.session.incrementTransactionNumber();
+  }
+
+  execute(server, callback) {
+    return this.model.execute(server, this.options, callback);
+  }
+}
+
+class CrudModel {
+  constructor(collection, options) {
+    options = Object.assign({}, options);
+    options = applyRetryableWrites(options, collection.s.db);
+    options = applyWriteConcern(options, { db: collection.s.db, collection }, options);
+    options.serializeFunctions = options.serializeFunctions || collection.s.serializeFunctions;
+
+    this.options = options;
+    this.namespace = collection.s.namespace;
+  }
+
+  execute(server, options, callback) {
+    if (!this.type || !this.operations) {
+      // TODO: Throw something?
+    }
+    server[this.type](this.namespace.toString(), this.operations, this.options, (err, result) => {
+      if (callback == null) return;
+      if (err) return handleCallback(callback, err, null);
+      if (result == null) return handleCallback(callback, null, null);
+      if (result.result.code) return handleCallback(callback, toError(result.result));
+      if (result.result.writeErrors)
+        return handleCallback(callback, toError(result.result.writeErrors[0]));
+      // Return the results
+      handleCallback(callback, null, result);
+    });
+  }
+}
+
+defineAspects(CrudWriteOperation, [Aspect.EXECUTE_WITH_SELECTION, Aspect.RETRY]);
+
+class BulkInsertModel extends CrudModel {
+  constructor(collection, options) {
+    super(collection, options);
+    this.type = INSERT;
+  }
+
+  canRetry() {
+    return true;
+  }
+}
+
+function prepareDocs(coll, docs, options) {
+  const forceServerObjectId =
+    typeof options.forceServerObjectId === 'boolean'
+      ? options.forceServerObjectId
+      : coll.s.db.options.forceServerObjectId;
+
+  // no need to modify the docs if server sets the ObjectId
+  if (forceServerObjectId === true) {
+    return docs;
+  }
+
+  return docs.map(doc => {
+    if (forceServerObjectId !== true && doc._id == null) {
+      doc._id = coll.s.pkFactory.createPk();
+    }
+
+    return doc;
+  });
+}
+
+function insertOneCallback(err, result, docs, callback) {
+  if (err) return handleCallback(callback, err);
+
+  // Workaround for pre 2.6 servers
+  if (result == null) return handleCallback(callback, null, { result: { ok: 1 } });
+
+  // Add docs to the list
+  result.ops = docs;
+  result.insertedCount = result.result.n;
+  result.insertedId = docs[0]._id;
+
+  // Return the results
+  handleCallback(callback, null, result);
+}
+
+class InsertOneModel extends BulkInsertModel {
+  constructor(args) {
```

Comment:
Is there a reason to pass in `args` here instead of `collection, doc, options` since we should always expect to have `collection`, `doc`, and `options`?

Reply:
Since we were approaching 4 or 5 parameters, my plan here was to eventually replace this with destructuring:

```js
constructor({ collection, doc, options }) {
}
```

That lets us be pretty flexible in adding additional, non-propagating parameters. A better example would be UpdateModel:

```js
constructor({ collection, selector, update, options, multi })
```

which is IMO neater.